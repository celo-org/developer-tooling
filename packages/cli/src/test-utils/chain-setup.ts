import { ContractKit } from '@celo/contractkit'
import { mineBlocks } from '@celo/dev-utils/lib/ganache-test'
import { addressToPublicKey } from '@celo/utils/lib/signatureUtils'
import BigNumber from 'bignumber.js'
import Web3 from 'web3'

export const GANACHE_EPOCH_SIZE = 100
export const MIN_LOCKED_CELO_VALUE = new BigNumber(Web3.utils.toWei('10000', 'ether')) // 10k CELO

const GROUP_COMMISION = new BigNumber(0.1)
const BLS_PUBLIC_KEY =
  '0x4fa3f67fc913878b068d1fa1cdddc54913d3bf988dbe5a36a20fa888f20d4894c408a6773f3d7bde11154f2a3076b700d345a42fd25a0e5e83f4db5586ac7979ac2053cd95d8f2efd3e959571ceccaa743e02cf4be3f5d7aaddb0b06fc9aff00'
const BLS_POP =
  '0xcdb77255037eb68897cd487fdd85388cbda448f617f874449d4b11588b0b7ad8ddc20d9bb450b513bb35664ea3923900'

export const registerAccount = async (kit: ContractKit, address: string) => {
  const accounts = await kit.contracts.getAccounts()

  if (!(await accounts.isAccount(address))) {
    await accounts.createAccount().sendAndWaitForReceipt({ from: address })
  }
}

export const registerAccountWithLockedGold = async (
  kit: ContractKit,
  address: string,
  value: string = MIN_LOCKED_CELO_VALUE.toFixed()
) => {
  await registerAccount(kit, address)

  const lockedGold = await kit.contracts.getLockedGold()

  await lockedGold.lock().sendAndWaitForReceipt({ from: address, value })
}

export const setupGroup = async (
  kit: ContractKit,
  groupAccount: string,
  groupCommission = GROUP_COMMISION
) => {
  await registerAccountWithLockedGold(
    kit,
    groupAccount,
    new BigNumber(MIN_LOCKED_CELO_VALUE).toFixed()
  )

  const validators = await kit.contracts.getValidators()

  await (
    await validators.registerValidatorGroup(groupCommission)
  ).sendAndWaitForReceipt({
    from: groupAccount,
  })
}

const setupValidator = async (kit: ContractKit, validatorAccount: string) => {
  await registerAccountWithLockedGold(kit, validatorAccount)

  const ecdsaPublicKey = await addressToPublicKey(validatorAccount, kit.connection.sign)
  const validators = await kit.contracts.getValidators()

  await validators
    .registerValidator(ecdsaPublicKey, BLS_PUBLIC_KEY, BLS_POP)
    .sendAndWaitForReceipt({
      from: validatorAccount,
    })
}

export const setupGroupAndAffiliateValidator = async (
  kit: ContractKit,
  groupAccount: string,
  validatorAccount: string
) => {
  await setupGroup(kit, groupAccount)
  await setupValidator(kit, validatorAccount)

  const validators = await kit.contracts.getValidators()

  await validators.affiliate(groupAccount).sendAndWaitForReceipt({ from: validatorAccount })

  await (
    await validators.addMember(groupAccount, validatorAccount)
  ).sendAndWaitForReceipt({
    from: groupAccount,
  })
}

export const voteForGroupFrom = async (
  kit: ContractKit,
  fromAddress: string,
  groupAddress: string,
  amount: BigNumber
) => {
  const election = await kit.contracts.getElection()

  await (await election.vote(groupAddress, amount)).sendAndWaitForReceipt({ from: fromAddress })
}

export const voteForGroupFromAndActivateVotes = async (
  kit: ContractKit,
  fromAddress: string,
  groupAddress: string,
  amount: BigNumber
) => {
  await voteForGroupFrom(kit, fromAddress, groupAddress, amount)

  await mineEpoch(kit)

  const election = await kit.contracts.getElection()

  const txos = await election.activate(fromAddress, false)

  await Promise.all(txos.map((txo) => txo.sendAndWaitForReceipt({ from: fromAddress })))
}

export const mineEpoch = async (kit: ContractKit) => {
  await mineBlocks(100, kit.web3)
}
