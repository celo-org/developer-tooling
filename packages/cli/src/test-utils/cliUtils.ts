import { Interfaces } from '@oclif/core'
import { BaseCommand } from '../base'

type AbstractConstructor<T> = new (...args: any[]) => T
interface Runner extends AbstractConstructor<BaseCommand> {
  run: typeof BaseCommand.run
}

export async function testLocally(
  command: Runner,
  argv: string[],
  config?: Interfaces.LoadOptions
) {
  const extendedArgv = [...argv, '--node', 'local']
  return command.run(extendedArgv, config)
}

// Removes font-formatting ANSI codes (colors/styles)
export const stripAnsiCodes = (text: string): string => {
  return text.replace(/\u001b\[.*?m/g, '')
}

export function stripAnsiCodesFromNestedArray(arrays: Array<string[]>) {
  return arrays.map((level0) => level0.map((level1) => stripAnsiCodes(level1)))
}

export const LONG_TIMEOUT_MS = 10 * 1000
export const EXTRA_LONG_TIMEOUT_MS = 20 * 1000
