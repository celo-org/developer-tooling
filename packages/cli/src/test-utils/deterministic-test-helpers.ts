import { HttpRpcCaller } from '@celo/connect'

/**
 * Mock gas prices to be deterministic across environments
 */
export const mockDeterministicGas = () => {
  return jest.spyOn(HttpRpcCaller.prototype, 'call').mockImplementation(async (method, _args) => {
    // Set deterministic gas prices that match CI environment
    if (method === 'eth_gasPrice') {
      return {
        result: '0x5d21dba00', // 25000000000 - deterministic gas price
        id: 1,
        jsonrpc: '2.0',
      }
    }
    if (method === 'eth_maxPriorityFeePerGas') {
      return {
        result: '0x4e3b29200', // 20000000000 - deterministic priority fee
        id: 1,
        jsonrpc: '2.0',
      }
    }
    if (method === 'eth_feeHistory') {
      return {
        result: {
          baseFeePerGas: ['0x5d21dba00'], // Same as gas price
          gasUsedRatio: [0.5],
          reward: [['0x4e3b29200']],
        },
        id: 1,
        jsonrpc: '2.0',
      }
    }
    // For other methods, call through to original implementation
    return HttpRpcCaller.prototype.call.call(this, method, _args)
  })
}

/**
 * Mock balance queries to return deterministic values
 */
export const mockDeterministicBalance = (expectedBalance: string) => {
  return jest.spyOn(HttpRpcCaller.prototype, 'call').mockImplementation(async (method, args) => {
    if (method === 'eth_getBalance') {
      return {
        result: expectedBalance, // Use the expected balance from CI snapshots
        id: 1,
        jsonrpc: '2.0',
      }
    }
    // For other methods, call through to original implementation
    return HttpRpcCaller.prototype.call.call(this, method, args)
  })
}