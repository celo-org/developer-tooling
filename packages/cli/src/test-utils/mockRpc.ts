import { HttpRpcCaller } from '@celo/connect'
import { RpcErrorCode } from 'viem'

export const mockRpc = () =>
  jest.spyOn(HttpRpcCaller.prototype, 'call').mockImplementation(async (method, _args) => {
    if (method === 'eth_maxPriorityFeePerGas') {
      return {
        result: '20000',
        id: 1,
        jsonrpc: '2.0',
      }
    }
    if (method === 'eth_gasPrice') {
      return {
        result: '30000',
        id: 1,
        jsonrpc: '2.0',
      }
    }
    return {
      result: 0,
      id: Math.random(),
      jsonrpc: '2.0',
    }
  })

const actualFetch = global.fetch
export const mockRpcFetch = ({
  method,
  result,
  error,
}: {
  method: string | string[]
  result?: any
  error?: { code: RpcErrorCode; message?: string }
}) => {
  const fetchMock = jest.fn(async (...args) => {
    if (args[1]?.body) {
      const body = JSON.parse(args[1].body.toString())
      if (typeof method === 'string' ? body.method === method : method.includes(body.method)) {
        return {
          ok: !error,
          headers: new Map([['Content-Type', 'application/json']]),
          json: () => {
            if (error) {
              return Promise.resolve({ error })
            } else {
              return Promise.resolve({ result, id: 1, jsonrpc: '2.0' })
            }
          },
        }
      }
    }
    // @ts-expect-error
    return actualFetch(...args)
  })
  // @ts-expect-error
  global.fetch = fetchMock

  return function restoreFetch() {
    global.fetch = actualFetch
  }
}
