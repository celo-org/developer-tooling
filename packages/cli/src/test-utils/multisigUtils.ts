import { multiSigABI, proxyABI } from '@celo/abis'
import { StrongAddress } from '@celo/base'
import { ContractKit } from '@celo/contractkit'
import { multiSigBytecode, proxyBytecode } from './constants'

export async function createMultisig(
  kit: ContractKit,
  owners: StrongAddress[],
  requiredSignatures: number,
  requiredInternalSignatures: number
): Promise<StrongAddress> {
  const accounts = (await kit.web3.eth.getAccounts()) as StrongAddress[]
  kit.defaultAccount = accounts[0]

  // Deploy Proxy contract
  const proxyDeploymentTx = await kit.sendTransaction({
    data: proxyBytecode,
  })
  const { contractAddress: proxyAddress } = await proxyDeploymentTx.waitReceipt()
  // Deploy MultiSig contract
  const multisigDeploymentTx = await kit.sendTransaction({
    data: multiSigBytecode,
  })
  const { contractAddress: multiSigAddress } = await multisigDeploymentTx.waitReceipt()

  // Configure and initialize MultiSig
  const initializerAbi = multiSigABI.find(
    (abi) => abi.type === 'function' && abi.name === 'initialize'
  )
  const proxy = new kit.web3.eth.Contract(proxyABI as any, proxyAddress)
  const baseFee = await kit.web3.eth.getBlock('latest').then((block: any) => block.baseFeePerGas)
  const priorityFee = kit.web3.utils.toWei('2', 'gwei')
  const initMethod = proxy.methods._setAndInitializeImplementation
  const callData = kit.web3.eth.abi.encodeFunctionCall(initializerAbi as any, [
    owners as any,
    requiredSignatures as any,
    requiredInternalSignatures as any,
  ])
  const initTx = initMethod(multiSigAddress, callData)
  await initTx.send({
    from: kit.defaultAccount,
    gas: await initTx.estimateGas({ from: kit.defaultAccount }),
    maxPriorityFeePerGas: priorityFee,
    maxFeePerGas: (BigInt(baseFee) + BigInt(priorityFee)).toString(),
  })
  const transferOwnershipMethod = proxy.methods._transferOwnership
  const changeOwnerTx = transferOwnershipMethod(proxyAddress)
  await changeOwnerTx.send({
    from: kit.defaultAccount,
    gas: await changeOwnerTx.estimateGas({ from: kit.defaultAccount }),
    maxPriorityFeePerGas: priorityFee,
    maxFeePerGas: (BigInt(baseFee) + BigInt(priorityFee)).toString(),
  })

  return proxyAddress as StrongAddress
}
