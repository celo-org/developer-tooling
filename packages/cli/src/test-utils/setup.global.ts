import baseSetup from '@celo/dev-utils/lib/ganache-setup'
// Has to import the matchers somewhere so that typescript knows the matchers have been made available
import _unused from '@celo/dev-utils/lib/matchers'
import * as path from 'path'

// Warning: There should be an unused import of '@celo/dev-utils/lib/matchers' above.
// If there is not, then your editor probably deleted it automatically.

export default async function globalSetup() {
  console.log('\nstarting ganache...')
  const chainDataPath = path.join(path.dirname(require.resolve('@celo/celo-devchain')), '../chains')
  // v X refers to core contract release X
  await baseSetup(path.resolve(chainDataPath), 'v11.tar.gz', {
    from_targz: true,
  })
  console.log('\n ganache started...')
  // it is necessary to disabled oclif integration with ts-node as
  // together it leads to a silent signit error and exit when tsconfk is loaded.
  // @ts-ignore - because global this doesnt have oclif property
  global.oclif = { tsnodeEnabled: false }
}
