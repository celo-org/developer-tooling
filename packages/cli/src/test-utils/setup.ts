jest.mock('@ledgerhq/hw-transport-node-hid', () => {
  return {
    default: {
      open: jest.fn(() => {
        return {
          send: jest.fn(() => new Promise(() => {})),
          decorateAppAPIMethods: jest.fn(),
          close: jest.fn(),
        }
      }),
    },
  }
})
