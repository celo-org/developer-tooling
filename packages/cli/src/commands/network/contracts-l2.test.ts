import { setupL2, testWithAnvil } from '@celo/dev-utils/lib/anvil-test'
import write from '@oclif/core/lib/cli-ux/write'
import { testLocallyWithWeb3Node } from '../../test-utils/cliUtils'
import Contracts from './contracts'
process.env.NO_SYNCCHECK = 'true'

testWithAnvil('network:contracts', (web3) => {
  test('runs', async () => {
    await setupL2(web3)
    const spy = jest.spyOn(write, 'stdout')
    await testLocallyWithWeb3Node(Contracts, ['--output', 'json'], web3)
    expect(spy.mock.calls).toMatchSnapshot()
  })
})
