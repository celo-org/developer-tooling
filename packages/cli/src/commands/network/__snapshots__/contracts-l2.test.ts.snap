// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`network:contracts runs 1`] = `
[
  [
    "[
  {
    "contract": "Accounts",
    "proxy": "0x8f24d37a4697e49AECd08d2b197E6968D2f007d3",
    "implementation": "0xCCf2141AC0F9a8bfD248d5b176d3a0FE6e92256C",
    "version": "1.1.5.0"
  },
  {
    "contract": "BlockchainParameters",
    "proxy": "0xFEB14C5787fd63a417612bF8Bc2232e169A7034F",
    "implementation": "0x1FCd4C3382702553316d810ea338E5eCCF176105",
    "version": "1.3.1.0"
  },
  {
    "contract": "DoubleSigningSlasher",
    "proxy": "0xA920a6fF4249f9A4CDdE3D67ff6275EdA6A1D89f",
    "implementation": "0xC5F6cEC0742589fD2bD88A5391A4376d70bc272E",
    "version": "1.1.2.0"
  },
  {
    "contract": "DowntimeSlasher",
    "proxy": "0xAADB35CAD2f922180106F50BF7ead66C5AD0f101",
    "implementation": "0x469E41Da18317E32C5BE85cDA387c6F5003Bc3Ab",
    "version": "2.0.1.0"
  },
  {
    "contract": "Election",
    "proxy": "0x5E386A280030a076528aD192BA94D7e31090F461",
    "implementation": "0x8a48Bcfe922e13F03958434a54Fe9D9956Ce0Aee",
    "version": "1.1.4.0"
  },
  {
    "contract": "EpochRewards",
    "proxy": "0xfff723D9B8f466Cd9e11BF4aaca171550aCF18Fc",
    "implementation": "0xB81c5ab7205c329A287c9123a83f55329BCEB8e9",
    "version": "1.1.2.0"
  },
  {
    "contract": "Escrow",
    "proxy": "0xd773882801F417427aE5C7A032296D93Fcf11DA9",
    "implementation": "0xAA1Ed790Ab56E34d04E2746C9DDBe9ea1E38825f",
    "version": "1.2.0.0"
  },
  {
    "contract": "FederatedAttestations",
    "proxy": "0x93f2E9307e3003a0a10A5171478CE18796aA2333",
    "implementation": "0x2C60D99fbb3D937865c129334aB2203aeDED69A9",
    "version": "1.1.0.0"
  },
  {
    "contract": "FeeCurrencyDirectory",
    "proxy": "0x42Fe5a2A61ed9705eb2F08a04A58CEB606D22f6a",
    "implementation": "0x63837e6E68C9070DF4f1CD70DCdcce4644b42b0E",
    "version": "1.1.0.0"
  },
  {
    "contract": "FeeCurrencyWhitelist",
    "proxy": "0x99f389e8a9903aF72ba481e8f000e8E229e529dA",
    "implementation": "0x84E366f7a6d4b8F9C00c86EB22de9f1Cbfff4327",
    "version": "NONE"
  },
  {
    "contract": "FeeHandler",
    "proxy": "0x04dE09026097347E2D540b6fbD9dC4aae3Ab0A90",
    "implementation": "0x8495b38627e5878ab84d1729df964F319caC7EEc",
    "version": "1.1.0.0"
  },
  {
    "contract": "Freezer",
    "proxy": "0x91dfD4C1B1262fad0f75A38D955B42b4bC586Bc1",
    "implementation": "0x871840D455493BA55F04056D557a92AB92c5901a",
    "version": "NONE"
  },
  {
    "contract": "GasPriceMinimum",
    "proxy": "0xb6b29b75e5616839eFc5883341042DAF72402966",
    "implementation": "0x1d1e8F4AA7B9d263830d68BBc8C82e5Fbf30Df7E",
    "version": "1.2.1.0"
  },
  {
    "contract": "GoldToken",
    "proxy": "0xfE8CbC1cFA1b3b8256f310bdfd40E60597083448",
    "implementation": "0x7E25F97124321c81EE39e80241fD2eDFAbe94999",
    "version": "1.1.2.0"
  },
  {
    "contract": "Governance",
    "proxy": "0xcF636ab1f15d3FB684A82cB1c69977DAfDFc6a0d",
    "implementation": "0x5889dd256a2bc829dDb44C4Fb1A140967854EB05",
    "version": "1.4.1.1"
  },
  {
    "contract": "LockedGold",
    "proxy": "0xc7F0E00681356896c06d5c810F0333ab30fBB8D1",
    "implementation": "0x017EfF77Eea3E83f50972FaCB6A585b46495f946",
    "version": "1.1.5.0"
  },
  {
    "contract": "MentoFeeHandlerSeller",
    "proxy": "0x677e4E735a36A7Ed935D424fCCe57a33831BF0Dc",
    "implementation": "0x944D551Ada142aa88388d6861b3c37fc99aB08dA",
    "version": "1.1.0.0"
  },
  {
    "contract": "OdisPayments",
    "proxy": "0xf7EfD92E61aB144483738B5F0f9cd46E6E9190BC",
    "implementation": "0x57B99AeE2b49D4D25f071c107bEcf20862B62b08",
    "version": "1.1.0.0"
  },
  {
    "contract": "Random",
    "proxy": "0x8982140ccFb38d7Dc439f953b37829f019A3e6e5",
    "implementation": "0xD2C3de3802a8130F4F171a7c1eebBb812B12Cf62",
    "version": "1.1.2.0"
  },
  {
    "contract": "Registry",
    "proxy": "0x000000000000000000000000000000000000ce10",
    "implementation": "0xF09E8DCA7ca021a11CC5927fC8fDD3b701Acf87e",
    "version": "NONE"
  },
  {
    "contract": "Reserve",
    "proxy": "0xe130B448fed06f3D5A2B196fAd4eE5b67f2E0805",
    "implementation": "0x4731729cA2F9BCAF57232BF2289b6BD59ea5F085",
    "version": "1.1.2.2"
  },
  {
    "contract": "SortedOracles",
    "proxy": "0xa2204011717369e044106e3bC93599E02538d65b",
    "implementation": "0xfDdDF67406A5E27a03bd3c78bF468f3D28D43f05",
    "version": "1.1.4.0"
  },
  {
    "contract": "StableToken",
    "proxy": "0xe6774BE4E5f97dB10cAFB4c00C74cFbdCDc434D9",
    "implementation": "0x8ccF94BE2763890dcf19e3c9B6c18558da565b22",
    "version": "NONE"
  },
  {
    "contract": "StableTokenBRL",
    "proxy": "0x2A3733dBc31980f02b12135C809b5da33BF3a1e9",
    "implementation": "0xAd798B7e15e484D01530a6413F15d26A368cD7CC",
    "version": "NONE"
  },
  {
    "contract": "StableTokenEUR",
    "proxy": "0xb7a33b4ad2B1f6b0a944232F5c71798d27Ad9272",
    "implementation": "0x589f510020b7fD6c7Ec94BDD352Cb6E8EB353c8A",
    "version": "NONE"
  },
  {
    "contract": "UniswapFeeHandlerSeller",
    "proxy": "0x27D282cBE154FD738de3242F80005cc2ca183981",
    "implementation": "0x88C21C7772521069611777747B5E140063A03B73",
    "version": "1.1.0.0"
  },
  {
    "contract": "Validators",
    "proxy": "0xA6aaDC309AA8E134d4a150eb4b58254801353FDc",
    "implementation": "0xD77fFbAd6F1aEc0F62AE4e46cc9156aB2EF1b1B1",
    "version": "1.3.0.0"
  }
]
",
  ],
]
`;
